File: CHAPT5/chapt5_ws/src/demo_python_tf/demo_python_tf/tf_listener.py
import rclpy
from rclpy.node import Node
import rclpy.time
from tf2_ros import TransformListener, Buffer
from tf_transformations import euler_from_quaternion
import math

class TFBroadcaster(Node): 
    def __init__(self):
        super().__init__('tf_listener')
        self.buffer = Buffer()
        self.listener = TransformListener(self.buffer, self)
        self.timer = self.create_timer(1.0, self.get_transform)
        
    def get_transform(self):
        # query
        try:
            result = self.buffer.lookup_transform('base_link', 'bottle_link', rclpy.time.Time(seconds=0.0), rclpy.time.Duration(seconds=1.0))
            transform = result.transform
            self.get_logger().info('translation: {}'.format(transform.translation))
            self.get_logger().info('rotation: {}'.format(transform.rotation))
            rotation_euler = euler_from_quaternion([transform.rotation.x, transform.rotation.y, transform.rotation.z, transform.rotation.w])
            self.get_logger().info('rotation RPY: {}'.format(rotation_euler))
        except Exception as e:
            self.get_logger().warn('get failed: {}'.format(str(e)))

        
def main():
    rclpy.init()
    node = TFBroadcaster()
    rclpy.spin(node)
    rclpy.shutdown()

File: CHAPT5/chapt5_ws/src/demo_cpp_tf/src/tf_listener.cpp
#include "geometry_msgs/msg/transform_stamped.hpp"
#include "rclcpp/rclcpp.hpp"
#include "tf2/LinearMath/Quaternion.h" //rf2 quaternion class
#include "tf2_geometry_msgs/tf2_geometry_msgs.h"
#include "tf2_ros/transform_listener.h"
#include "tf2_ros/buffer.h"
#include "tf2/utils.h"
#include <chrono>
using namespace std::chrono_literals;

class TFListener: public rclcpp::Node{
public:
    TFListener(): Node("tf_broadcast"){
        buffer = std::make_shared<tf2_ros::Buffer>(this->get_clock());
        listener = std::make_shared<tf2_ros::TransformListener>(*buffer, this);
        timer = this->create_wall_timer(1s, std::bind(&TFListener::getTransform, this));
    }

    void getTransform(){
        try{
            const auto transform = buffer->lookupTransform("base_link", "target_point", this->get_clock()->now(), rclcpp::Duration::from_seconds(1));
            
            auto translation = transform.transform.translation;
            auto rotation = transform.transform.rotation;
            double y, p, r;
            tf2::getEulerYPR(rotation, y, p, r);
            RCLCPP_INFO(get_logger(), "translation: %f, %f %f", translation.x, translation.y, translation.z);
            RCLCPP_INFO(get_logger(), "rotation: %f, %f %f", y, p, r);

        }
        catch(const std::exception& e){
            RCLCPP_WARN(get_logger(), "%s", e.what());
        }        
    }


private:
    std::shared_ptr<tf2_ros::TransformListener> listener;

    rclcpp::TimerBase::SharedPtr timer;

    std::shared_ptr<tf2_ros::Buffer> buffer;
};

int main(int argc, char** argv){
    rclcpp::init(argc, argv);
    auto node = std::make_shared<TFListener>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}


